clip.md

```diff
@@ -0,0 +1,443 @@
+      <clip-tip> CLIP_TIP </clip-tip>
+      <clip-chip> CLIP_CHIP </clip-chip>
+      <current-time> NOW </current-time>
+
+      requires CLIP_STOPPED <ClipStop noNewKick
+      andBool TAB >Rad rad(0)
+      andBool LOT >Wad wad(0)
+      andBool USR =/=K 0:Address
+      andBool KICKS +Int 1 >Int 0
+      andBool ( ( ( Wad2Ray(VALUE) ) /Ray PAR )  *Ray CLIP_BUF ) >Ray ray(0)
+```
+
+```k
+endmodule
+```
+
+A PRE module was introduced to avoid circular imports.
+
+```k
+module CLIP
+    import DOG
+    import PRE-CLIP
+```
+
+```k
+   syntax ClipStep ::= "upchost"
+// -------------------------
+   rule <k> Clip . upchost => . ... </k>
+      <clip-chost> _ => DUST *Rad CHOP </clip-chost>
+      <clip-ilk> CLIP_ILK </clip-ilk>
+      <vat-ilks> ... CLIP_ILK |-> Ilk( ... dust: DUST) ... </vat-ilks>
+      <dog-ilks> ... CLIP_ILK |-> Ilk( ... chop: CHOP) ... </dog-ilks>
+```
+
+
+```k
+   syntax ClipLockStep ::= "redo" Int Address
+// ------------------------------------------
+   rule <k> Clip . redo ID KPR
+   => #fun(PRICE
+   => #fun(DONE
+   => #fun(TOP_FINAL
+   => #fun(COIN
+   => ( #if ( CLIP_TIP >Rad rad(0) orBool CLIP_CHIP >Wad wad(0) ) #then ( #if ( TAB >=Rad CLIP_CHOST andBool ( Wad2Ray(LOT) *Ray ( Wad2Ray(VALUE) /Ray SPOT_PAR ) ) >=Ray CLIP_CHOST ) #then call CLIP_VAT . suck CLIP_VOW KPR COIN #else . #fi ) #else . #fi )
+   ~> DONE
+   ~> emitRedo ID TOP_FINAL TAB LOT USR KPR COIN
+   ) ( #if ( CLIP_TIP >Rad rad(0) orBool CLIP_CHIP >Wad wad(0) )  #then ( #if ( TAB >=Rad CLIP_CHOST andBool ( Wad2Ray(LOT) *Ray ( Wad2Ray(VALUE) /Ray SPOT_PAR ) ) >=Ray CLIP_CHOST ) #then CLIP_TIP +Rad ( Rad2Wad(TAB) *Wad CLIP_CHIP ) #else wad(0) #fi ) #else wad(0) #fi )
+   )( ( Wad2Ray(VALUE) /Ray SPOT_PAR ) *Ray CLIP_BUF )
+   )( ( (NOW -Int TIC ) >Int CLIP_TAIL ) orBool ( (PRICE /Ray TOP) <Ray CLIP_CUSP ) )
+   )( Abacus CLIP_ABACUS . price TOP (NOW -Int TIC) )
+   ... </k>
+   <clip-abacus> CLIP_ABACUS </clip-abacus>
+   <clip-buf> CLIP_BUF </clip-buf>
+   <clip-chip> CLIP_CHIP </clip-chip>
+   <clip-chost> CLIP_CHOST </clip-chost>
+   <clip-cusp> CLIP_CUSP </clip-cusp>
+   <clip-ilk> CLIP_ILK </clip-ilk>
+   <clip-tail> CLIP_TAIL </clip-tail>
+   <clip-tip> CLIP_TIP </clip-tip>
+   <clip-sales> ... ID |-> ClipSale( ... tab: TAB, lot: LOT, usr: USR, tic: (TIC => NOW),top: (TOP => ( ( Wad2Ray(VALUE) /Ray SPOT_PAR ) *Ray CLIP_BUF ) ) ) ... </clip-sales>
+   <clip-vat> CLIP_VAT </clip-vat>
+   <clip-vow> CLIP_VOW </clip-vow>
+   <current-time> NOW </current-time>
+   <spot-ilks> CLIP_ILK |-> SpotIlk(... pip: VALUE ) </spot-ilks>
+   <spot-par> SPOT_PAR </spot-par>
+   <clip-stopped> CLIP_STOPPED </clip-stopped>
+   requires CLIP_STOPPED <ClipStop noNewKickOrRedo
+   andBool USR =/=K 0:Address
+   andBool ( ( Wad2Ray(VALUE) /Ray SPOT_PAR ) *Ray CLIP_BUF ) >Ray ray(0) //top > 0
+
+   rule <k> DONE:Bool ~> emitRedo ID TOP_FINAL TAB LOT USR KPR COIN => emitRedo ID TOP_FINAL TAB LOT USR KPR COIN ... </k>
+      requires DONE
+```
+
+TODO add external call
```

Comment:
Will this be added to this PR, or is it being saved for later?

Reply:
It will be added to this PR but at the time I was unsure of how to model an external call in this framework, will talk with Everett about this